filters/include/pcl/filters/impl/crop_hull.hpp

```diff
@@ -113,16 +102,44 @@ pcl::CropHull<PointT>::getHullCloudRange ()
       if (pt[i] > cloud_max[i]) cloud_max[i] = pt[i];
     }
   }
-  
+
   return (cloud_max - cloud_min);
 }
 
 //////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 template<typename PointT> template<unsigned PlaneDim1, unsigned PlaneDim2> void 
 pcl::CropHull<PointT>::applyFilter2D (PointCloud &output)
 {
+  std::vector<int> indices;
+  applyFilter2D<PlaneDim1, PlaneDim2>(indices);
+  pcl::copyPointCloud(*input_, indices, output);
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+template<typename PointT> template<unsigned PlaneDim1, unsigned PlaneDim2> void 
+pcl::CropHull<PointT>::applyFilter2D (std::vector<int> &indices)
+{
+  Eigen::Vector4f const minPt = crop_box_.getMin();
+  Eigen::Vector4f const maxPt = crop_box_.getMax();
+  std::vector<bool> cropInlierMask(input_->size(), true);
+  std::vector<int> cropBoxRemovedIndices;
+  for (std::size_t i = 0; i < indices_->size(); ++i)
+  {
+    int const idx = (*indices_)[i];
+    Eigen::Vector3f pt = input_->points[idx].getVector3fMap();
+    if (pt[PlaneDim1] < minPt[PlaneDim1] || pt[PlaneDim1] > maxPt[PlaneDim2] ||
+        pt[PlaneDim2] < minPt[PlaneDim2] || pt[PlaneDim2] > maxPt[PlaneDim2])
+    {
+      cropInlierMask.at(idx) = false;
+      cropBoxRemovedIndices.push_back(idx);
+    }
+  }
+
   for (std::size_t index = 0; index < indices_->size (); index++)
   {
+    if (!cropInlierMask[(*indices_)[index]]) {
+      continue;
+    }
     // iterate over polygons faster than points because we expect this data
     // to be, in general, more cache-local - the point cloud might be huge
     std::size_t poly;
```

Comment:
Using `find_if` and the comparison trick by @Morwenn can simplify the code for L145-L164:
```cpp
    const auto poly_idx = std::find_if(hull_polygons_.cbegin(), hull_polygons_.cend(),
      [](const auto& poly) {
        return isPointIn2DPolyWithVertIndices<PlaneDim1,PlaneDim2> (
          input_->points[(*indices_)[index]], hull_polygons_[poly], *hull_cloud_);
      });
    const bool not_found_in_polygons = (poly_idx == hull_polygons_.cend());
    if (crop_outside_ == not_found_in_polygons) {
      // valid index: either found inside and no need to crop outside
      //                     or not found inside and need to crop outside
      indices.push_back ((*indices_)[index]);
    }
```

Reply:
applied the suggestion